File: Code/OfRectanglesAndSquares/OfRectanglesAndSquares/CalculatorWithOptimization.cs
namespace OfRectanglesAndSquares
{
    /// <summary>
    /// This class has an optimized calculus algorithm
    /// </summary>
    public class CalculatorWithOptimization : Calculator
    {
        public override double CalcArea(Rectangle r)
        {
            if (r.GetWidth() == 0.0) return 0.0;
            if (r.GetHeight() == 0.0) return 0.0;

            return base.CalcArea(r);
        }
    }
}

File: Code/OfRectanglesAndSquares.Java/src/OfRectanglesAndSquares/CalculatorWithOptimization.java
package OfRectanglesAndSquares;

/**
 * This class has an optimized calculus algorithm only for rectangles
 */
public class CalculatorWithOptimization extends Calculator {
    
	@Override
    public double calcArea(IRectangle r) {
        if (r.getWidth() == 0.0) return 0.0;
        if (r.getHeight() == 0.0) return 0.0;

        return super.calcArea(r);
    }

}
